pkg/controllers/cassandra/pilot/pilot.go

```diff
@@ -0,0 +1,165 @@
+package pilot
+
+import (
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	navigator "github.com/jetstack/navigator/pkg/client/clientset/versioned"
+	navlisters "github.com/jetstack/navigator/pkg/client/listers/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/cassandra/util"
+	"k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	appslisters "k8s.io/client-go/listers/apps/v1beta1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+type Interface interface {
+	Sync(*v1alpha1.CassandraCluster) error
+}
+
+type pilotControl struct {
+	naviClient   navigator.Interface
+	pilots       navlisters.PilotLister
+	pods         corelisters.PodLister
+	statefulSets appslisters.StatefulSetLister
+	recorder     record.EventRecorder
+}
+
+var _ Interface = &pilotControl{}
+
+func NewControl(
+	naviClient navigator.Interface,
+	pilots navlisters.PilotLister,
+	pods corelisters.PodLister,
+	statefulSets appslisters.StatefulSetLister,
+	recorder record.EventRecorder,
+) *pilotControl {
+	return &pilotControl{
+		naviClient:   naviClient,
+		pilots:       pilots,
+		pods:         pods,
+		statefulSets: statefulSets,
+		recorder:     recorder,
+	}
+
+}
+
+func (c *pilotControl) clusterPods(cluster *v1alpha1.CassandraCluster) ([]*v1.Pod, error) {
+	var clusterPods []*v1.Pod
+	allPods, err := c.pods.Pods(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return clusterPods, err
+	}
+	for _, pod := range allPods {
+		podControlledByCluster, err := controllers.PodControlledByCluster(
+			cluster,
+			pod,
+			c.statefulSets,
+		)
+		if err != nil {
+			return clusterPods, err
+		}
+		if !podControlledByCluster {
+			continue
+		}
+		clusterPods = append(clusterPods, pod)
+	}
+	return clusterPods, nil
+}
+
+func (c *pilotControl) createOrUpdatePilot(cluster *v1alpha1.CassandraCluster, pod *v1.Pod) error {
+	desiredPilot := PilotForCluster(cluster, pod)
+	client := c.naviClient.NavigatorV1alpha1().Pilots(desiredPilot.GetNamespace())
+	lister := c.pilots.Pilots(desiredPilot.GetNamespace())
+	existingPilot, err := lister.Get(desiredPilot.GetName())
+	if k8sErrors.IsNotFound(err) {
+		_, err = client.Create(desiredPilot)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	err = util.OwnerCheck(existingPilot, cluster)
+	if err != nil {
+		return err
+	}
+	desiredPilot = existingPilot.DeepCopy()
+	updatePilotForCluster(cluster, pod, desiredPilot)
+	_, err = client.Update(desiredPilot)
```

Comment:
So from what I can see here, this will cause Navigator to go into a sync loop, continuously updating the Pilot resource. We don't ever check if the Pilot is already reconciled (e.g. it's spec is up to date), so we end out updating to contain the same spec as is already there. This will still however increment the ResourceVersion, thus causing this loop to be triggered once again (and so on).

There are a couple of ways (that I can think of) to deal with this:

* Perform a `reflect.DeepEqual` on the `pilot.Spec` before calling Update. This has the downside of being inefficient (DeepEqual is an expensive call), however it is accurate and easy to do.

* Write a hash of the spec into the annotations of the Pilot, so that we can quickly compare the old and new hashes.

* Manually compare each field, but this is tedious and error prone.

Can you think of anything else here?

Reply:
Additionally, we need to update the strategy.go file for Cassandra resources to not perform updates to the `cassandracluster.status` block when plain Update is called (and instead, `UpdateStatus` should be used if the Status block needs updating).

Right now, each time Update is called we will wipe out all fields in Status, causing Pilots and Navigator to fight with each other indefinitely (causing more loops)

Reply:
> Write a hash of the spec into the annotations of the Pilot, so that we can quickly compare the old and new hashes.

I've gone with that option. I found the ES hashing code and adapted that. Also added labels to the hash since the controller updates those too.

Reply:
> Right now, each time Update is called we will wipe out all fields in Status, causing Pilots and Navigator to fight with each other indefinitely (causing more loops)

I don't think that's the case. I'm taking the latest Pilot from the lister and then updating labels and in future, spec Not replacing Pilot.Status.